File: PHP/One/Transforms/InferTypesPlugins/ResolveFieldAndPropertyAccess.php
<?php

namespace One\Transforms\InferTypesPlugins\ResolveFieldAndPropertyAccess;

use One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use One\Ast\Expressions\Expression;
use One\Ast\Expressions\PropertyAccessExpression;
use One\Ast\Expressions\UnresolvedCallExpression;
use One\Ast\References\ClassReference;
use One\Ast\References\Reference;
use One\Ast\References\StaticFieldReference;
use One\Ast\References\StaticPropertyReference;
use One\Ast\References\InstanceFieldReference;
use One\Ast\References\InstancePropertyReference;
use One\Ast\References\ThisReference;
use One\Ast\References\SuperReference;
use One\Ast\References\EnumReference;
use One\Ast\References\StaticThisReference;
use One\Ast\Types\Class_;
use One\Ast\Types\Method;
use One\Ast\Types\Interface_;
use One\Ast\AstTypes\ClassType;
use One\Ast\AstTypes\InterfaceType;
use One\Ast\AstTypes\AnyType;
use One\Ast\AstTypes\TypeHelper;
use One\Transforms\InferTypesPlugins\Helpers\GenericsResolver\GenericsResolver;

class ResolveFieldAndPropertyAccess extends InferTypesPlugin {
    function __construct() {
        parent::__construct("ResolveFieldAndPropertyAccess");
        
    }
    
    protected function getStaticRef($cls, $memberName) {
        $field = \OneLang\ArrayHelper::find($cls->fields, function ($x) use ($memberName) { return $x->name === $memberName; });
        if ($field !== null && $field->isStatic)
            return new StaticFieldReference($field);
        
        $prop = \OneLang\ArrayHelper::find($cls->properties, function ($x) use ($memberName) { return $x->name === $memberName; });
        if ($prop !== null && $prop->isStatic)
            return new StaticPropertyReference($prop);
        
        $this->errorMan->throw("Could not resolve static member access of a class: " . $cls->name . "::" . $memberName);
        return null;
    }
    
    protected function getInstanceRef($cls, $memberName, $obj) {
        while (true) {
            $field = \OneLang\ArrayHelper::find($cls->fields, function ($x) use ($memberName) { return $x->name === $memberName; });
            if ($field !== null && !$field->isStatic)
                return new InstanceFieldReference($obj, $field);
            
            $prop = \OneLang\ArrayHelper::find($cls->properties, function ($x) use ($memberName) { return $x->name === $memberName; });
            if ($prop !== null && !$prop->isStatic)
                return new InstancePropertyReference($obj, $prop);
            
            if ($cls->baseClass === null)
                break;
            
            $cls = ($cls->baseClass)->decl;
        }
        
        $this->errorMan->throw("Could not resolve instance member access of a class: " . $cls->name . "::" . $memberName);
        return null;
    }
    
    protected function getInterfaceRef($intf, $memberName, $obj) {
        $field = \OneLang\ArrayHelper::find($intf->fields, function ($x) use ($memberName) { return $x->name === $memberName; });
        if ($field !== null && !$field->isStatic)
            return new InstanceFieldReference($obj, $field);
        
        foreach ($intf->baseInterfaces as $baseIntf) {
            $res = $this->getInterfaceRef(($baseIntf)->decl, $memberName, $obj);
            if ($res !== null)
                return $res;
        }
        return null;
    }
    
    protected function transformPA($expr) {
        if ($expr->object instanceof ClassReference)
            return $this->getStaticRef($expr->object->decl, $expr->propertyName);
        
        if ($expr->object instanceof StaticThisReference)
            return $this->getStaticRef($expr->object->cls, $expr->propertyName);
        
        $expr->object = $this->main->runPluginsOn($expr->object);
        
        if ($expr->object instanceof ThisReference)
            return $this->getInstanceRef($expr->object->cls, $expr->propertyName, $expr->object);
        
        $type = $expr->object->getType();
        if ($type instanceof ClassType)
            return $this->getInstanceRef($type->decl, $expr->propertyName, $expr->object);
        else if ($type instanceof InterfaceType) {
            $ref = $this->getInterfaceRef($type->decl, $expr->propertyName, $expr->object);
            if ($ref === null)
                $this->errorMan->throw("Could not resolve instance member access of a interface: " . $type->repr() . "::" . $expr->propertyName);
            return $ref;
        }
        else if ($type === null)
            $this->errorMan->throw("Type was not inferred yet (prop=\"" . $expr->propertyName . "\")");
        else if ($type instanceof AnyType)
            //this.errorMan.throw(`Object has any type (prop="${expr.propertyName}")`);
            $expr->setActualType(AnyType::$instance);
        else
            $this->errorMan->throw("Expected class as variable type, but got: " . $type->repr() . " (prop=\"" . $expr->propertyName . "\")");
        
        return $expr;
    }
    
    function canTransform($expr) {
        return $expr instanceof PropertyAccessExpression && !($expr->object instanceof EnumReference) && !($expr->parentNode instanceof UnresolvedCallExpression && $expr->parentNode->func === $expr) && !($expr->actualType instanceof AnyType);
    }
    
    function transform($expr) {
        return $this->transformPA($expr);
    }
    
    function canDetectType($expr) {
        return $expr instanceof InstanceFieldReference || $expr instanceof InstancePropertyReference || $expr instanceof StaticFieldReference || $expr instanceof StaticPropertyReference;
    }
    
    function detectType($expr) {
        if ($expr instanceof InstanceFieldReference) {
            $actualType = GenericsResolver::fromObject($expr->object)->resolveType($expr->field->type, true);
            $expr->setActualType($actualType, false, TypeHelper::isGeneric($expr->object->actualType));
            return true;
        }
        else if ($expr instanceof InstancePropertyReference) {
            $actualType = GenericsResolver::fromObject($expr->object)->resolveType($expr->property->type, true);
            $expr->setActualType($actualType);
            return true;
        }
        else if ($expr instanceof StaticPropertyReference) {
            $expr->setActualType($expr->decl->type, false, false);
            return true;
        }
        else if ($expr instanceof StaticFieldReference) {
            $expr->setActualType($expr->decl->type, false, false);
            return true;
        }
        
        return false;
    }
}

File: CSharp/One/Transforms/InferTypesPlugins/ResolveFieldAndPropertyAccess.cs
using One.Ast;
using One.Transforms.InferTypesPlugins.Helpers;

namespace One.Transforms.InferTypesPlugins
{
    public class ResolveFieldAndPropertyAccess : InferTypesPlugin {
        public ResolveFieldAndPropertyAccess(): base("ResolveFieldAndPropertyAccess")
        {
            
        }
        
        protected Reference getStaticRef(Class cls, string memberName)
        {
            var field = cls.fields.find(x => x.name == memberName);
            if (field != null && field.isStatic)
                return new StaticFieldReference(field);
            
            var prop = cls.properties.find(x => x.name == memberName);
            if (prop != null && prop.isStatic)
                return new StaticPropertyReference(prop);
            
            this.errorMan.throw_($"Could not resolve static member access of a class: {cls.name}::{memberName}");
            return null;
        }
        
        protected Reference getInstanceRef(Class cls, string memberName, Expression obj)
        {
            while (true) {
                var field = cls.fields.find(x => x.name == memberName);
                if (field != null && !field.isStatic)
                    return new InstanceFieldReference(obj, field);
                
                var prop = cls.properties.find(x => x.name == memberName);
                if (prop != null && !prop.isStatic)
                    return new InstancePropertyReference(obj, prop);
                
                if (cls.baseClass == null)
                    break;
                
                cls = (((ClassType)cls.baseClass)).decl;
            }
            
            this.errorMan.throw_($"Could not resolve instance member access of a class: {cls.name}::{memberName}");
            return null;
        }
        
        protected Reference getInterfaceRef(Interface intf, string memberName, Expression obj)
        {
            var field = intf.fields.find(x => x.name == memberName);
            if (field != null && !field.isStatic)
                return new InstanceFieldReference(obj, field);
            
            foreach (var baseIntf in intf.baseInterfaces) {
                var res = this.getInterfaceRef((((InterfaceType)baseIntf)).decl, memberName, obj);
                if (res != null)
                    return res;
            }
            return null;
        }
        
        protected Expression transformPA(PropertyAccessExpression expr)
        {
            if (expr.object_ is ClassReference classRef)
                return this.getStaticRef(classRef.decl, expr.propertyName);
            
            if (expr.object_ is StaticThisReference statThisRef)
                return this.getStaticRef(statThisRef.cls, expr.propertyName);
            
            expr.object_ = this.main.runPluginsOn(expr.object_);
            
            if (expr.object_ is ThisReference thisRef)
                return this.getInstanceRef(thisRef.cls, expr.propertyName, thisRef);
            
            var type = expr.object_.getType();
            if (type is ClassType classType)
                return this.getInstanceRef(classType.decl, expr.propertyName, expr.object_);
            else if (type is InterfaceType intType) {
                var ref_ = this.getInterfaceRef(intType.decl, expr.propertyName, expr.object_);
                if (ref_ == null)
                    this.errorMan.throw_($"Could not resolve instance member access of a interface: {intType.repr()}::{expr.propertyName}");
                return ref_;
            }
            else if (type == null)
                this.errorMan.throw_($"Type was not inferred yet (prop=\"{expr.propertyName}\")");
            else if (type is AnyType)
                //this.errorMan.throw(`Object has any type (prop="${expr.propertyName}")`);
                expr.setActualType(AnyType.instance);
            else
                this.errorMan.throw_($"Expected class as variable type, but got: {type.repr()} (prop=\"{expr.propertyName}\")");
            
            return expr;
        }
        
        public override bool canTransform(Expression expr)
        {
            return expr is PropertyAccessExpression propAccExpr && !(propAccExpr.object_ is EnumReference) && !(propAccExpr.parentNode is UnresolvedCallExpression unrCallExpr && unrCallExpr.func == propAccExpr) && !(propAccExpr.actualType is AnyType);
        }
        
        public override Expression transform(Expression expr)
        {
            return this.transformPA(((PropertyAccessExpression)expr));
        }
        
        public override bool canDetectType(Expression expr)
        {
            return expr is InstanceFieldReference instFieldRef || expr is InstancePropertyReference || expr is StaticFieldReference || expr is StaticPropertyReference;
        }
        
        public override bool detectType(Expression expr)
        {
            if (expr is InstanceFieldReference instFieldRef2) {
                var actualType = GenericsResolver.fromObject(instFieldRef2.object_).resolveType(instFieldRef2.field.type, true);
                instFieldRef2.setActualType(actualType, false, TypeHelper.isGeneric(instFieldRef2.object_.actualType));
                return true;
            }
            else if (expr is InstancePropertyReference instPropRef) {
                var actualType = GenericsResolver.fromObject(instPropRef.object_).resolveType(instPropRef.property.type, true);
                instPropRef.setActualType(actualType);
                return true;
            }
            else if (expr is StaticPropertyReference statPropRef) {
                statPropRef.setActualType(statPropRef.decl.type, false, false);
                return true;
            }
            else if (expr is StaticFieldReference statFieldRef) {
                statFieldRef.setActualType(statFieldRef.decl.type, false, false);
                return true;
            }
            
            return false;
        }
    }
}